import_obfuscation.go

```diff
@@ -0,0 +1,535 @@
+					// that means garbling strings might effect the behavior of the compiled binary
+					if s.Kind == goobj2.SRODATA && s.Data != nil && !strings.HasPrefix(s.Name, "go.string.") {
+						var dataTyp dataType
+						if strings.HasPrefix(s.Name, "type..importpath.") {
+							dataTyp = importPath
+						} else if strings.HasPrefix(s.Name, "type..namedata.") {
+							dataTyp = namedata
+						}
+						s.Data = garbleSymData(s.Data, privateImports, garbledImports, dataTyp, &buf)
+
+						if s.Size != 0 {
+							s.Size = uint32(len(s.Data))
+						}
+					}
+					s.Name = garbleSymbolName(s.Name, privateImports, garbledImports, &sb)
+
+					for i := range s.Reloc {
+						s.Reloc[i].Name = garbleSymbolName(s.Reloc[i].Name, privateImports, garbledImports, &sb)
+					}
+					if s.Type != nil {
+						s.Type.Name = garbleSymbolName(s.Type.Name, privateImports, garbledImports, &sb)
+					}
+					if s.Func != nil {
+						for i := range s.Func.FuncData {
+							s.Func.FuncData[i].Sym.Name = garbleSymbolName(s.Func.FuncData[i].Sym.Name, privateImports, garbledImports, &sb)
+						}
+						for _, inl := range s.Func.InlTree {
+							inl.Func.Name = garbleSymbolName(inl.Func.Name, privateImports, garbledImports, &sb)
+						}
+
+						// remove unneeded debug aux symbols
+						s.Func.DwarfInfo = nil
+						s.Func.DwarfLoc = nil
+						s.Func.DwarfRanges = nil
+						s.Func.DwarfDebugLines = nil
+					}
+				}
+			}
+			for i := range am.SymRefs {
+				am.SymRefs[i].Name = garbleSymbolName(am.SymRefs[i].Name, privateImports, garbledImports, &sb)
+			}
+		}
+
+		if err := p.pkg.Write(p.path); err != nil {
+			return nil, fmt.Errorf("error writing objfile %s at %s: %v", p.pkg.ImportPath, p.path, err)
+		}
+	}
+
+	// garble importcfg so the linker knows where to find garbled imports
+	newCfg, err := os.Create(importCfgPath)
+	if err != nil {
+		return nil, fmt.Errorf("error creating importcfg: %v", err)
+	}
+	defer newCfg.Close()
+	newCfgWr := bufio.NewWriter(newCfg)
+
+	for pkgPath, info := range importCfg {
+		if isPrivate(pkgPath) {
+			pkgPath = hashImport(pkgPath, garbledImports)
+		}
+		if info.IsSharedLib {
+			newCfgWr.WriteString("packageshlib")
+		} else {
+			newCfgWr.WriteString("packagefile")
+		}
+
+		newCfgWr.WriteRune(' ')
+		newCfgWr.WriteString(pkgPath)
+		newCfgWr.WriteRune('=')
+		newCfgWr.WriteString(info.Path)
+		newCfgWr.WriteRune('\n')
+	}
+
+	if err := newCfgWr.Flush(); err != nil {
+		return nil, fmt.Errorf("error writing importcfg: %v", err)
+	}
+
+	return garbledImports, nil
+}
+
+// importPathCombos returns a list of import paths that
+// could all potentially be in symbol names of the
+// package that imported 'path'.
+// TODO: last element returned should get same buildID
+// as full path?
+// ie github.com/foo/bar.buildID == bar.buildID
+func importPathCombos(path string) []string {
+	paths := strings.Split(path, "/")
+	combos := make([]string, 0, len(paths))
+
+	var restPrivate bool
+	if isPrivate(paths[0]) {
```

Comment:
is this right? because if I have `path := "foo/bar/baz"` and `GOPRIVATE=foo/bar`, you'll incorrectly think that the path is not private because you're just doing `isPrivate("foo")`.

Reply:
Yep, that's correct. Basically what I'm trying to do there is build the shortest import path that is in GOPRIVATE, then add private import paths and names from that.